boa/src/builtins/value/operations.rs

```diff
@@ -44,32 +53,129 @@ impl Value {
+                Some(ref a) => a == b,
+                None => false,
+            },
+
+            // 8. If Type(x) is Boolean, return the result of the comparison ! ToNumber(x) == y.
             (ValueData::Boolean(_), _) => {
                 other.equals(&mut Value::from(self.to_integer()), interpreter)
             }
+
+            // 9. If Type(y) is Boolean, return the result of the comparison x == ! ToNumber(y).
             (_, ValueData::Boolean(_)) => {
                 self.equals(&mut Value::from(other.to_integer()), interpreter)
             }
+
+            // 10. If Type(x) is either String, Number, BigInt, or Symbol and Type(y) is Object, return the result
+            // of the comparison x == ? ToPrimitive(y).
             (ValueData::Object(_), _) => {
                 let mut primitive = interpreter.to_primitive(self, None);
                 primitive.equals(other, interpreter)
             }
+
+            // 11. If Type(x) is Object and Type(y) is either String, Number, BigInt, or Symbol, return the result
+            // of the comparison ? ToPrimitive(x) == y.
             (_, ValueData::Object(_)) => {
                 let mut primitive = interpreter.to_primitive(other, None);
                 primitive.equals(self, interpreter)
             }
+
+            // 12. If Type(x) is BigInt and Type(y) is Number, or if Type(x) is Number and Type(y) is BigInt, then
+            //    a. If x or y are any of NaN, +∞, or -∞, return false.
+            //    b. If the mathematical value of x is equal to the mathematical value of y, return true; otherwise return false.
+            (ValueData::BigInt(ref a), ValueData::Rational(ref b)) => a == b,
+            (ValueData::Rational(ref a), ValueData::BigInt(ref b)) => a == b,
+            (ValueData::BigInt(ref a), ValueData::Integer(ref b)) => a == b,
+            (ValueData::Integer(ref a), ValueData::BigInt(ref b)) => a == b,
+
+            // 13. Return false.
             _ => false,
         }
     }
 }
 
+/// This function takes a string and conversts it to BigInt type.
+///
+/// If the result is `NaN` than `None` is returned.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-stringtobigint
+pub fn string_to_bigint(string: &str) -> Option<BigInt> {
+    if string.is_empty() {
+        return Some(BigInt::from(0));
+    }
+
+    BigInt::from_str(string).ok()
+}
+
+/// The internal comparison abstract operation SameValue(x, y),
+/// where x and y are ECMAScript language values, produces true or false.
+/// Such a comparison is performed as follows:
+///
+/// https://tc39.es/ecma262/#sec-samevalue
+/// strict mode currently compares the pointers
+pub fn same_value(x: &Value, y: &Value, strict: bool) -> bool {
+    if strict {
+        // Do both Values point to the same underlying valueData?
+        return std::ptr::eq(x.data(), y.data());
```

Comment:
Nice!
Where does `data()` come from? 
I didn't see it in Gc's documentation

Reply:
Also was there a reason you didn't use https://docs.rs/gc/0.3.4/gc/struct.Gc.html#method.ptr_eq ?

Reply:
> Where does `data()` come from?
> I didn't see it in Gc's documentation

I abstracted Value as a `struct Value(Gc<ValueData>)` and added a helper function `Value::data()` which returns a `&ValueData` in #383 instead of doing `&**x` you can write just `x.data()`.

> Also was there a reason you didn't use [docs.rs/gc/0.3.4/gc/struct.Gc.html#method.ptr_eq](https://docs.rs/gc/0.3.4/gc/struct.Gc.html#method.ptr_eq) ?

Because `Value` is abstracted as a `struct` you can not pass is as a `Gc<ValueData>`. This functionally does the same thing